File: hedera-mirror-common/src/test/java/com/hedera/mirror/common/domain/topic/TopicMessageTest.java
/*
 * Copyright (C) 2020-2025 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.mirror.common.domain.topic;

import static com.hedera.mirror.common.converter.ObjectToStringSerializer.OBJECT_MAPPER;
import static org.assertj.core.api.Assertions.assertThat;

import com.hedera.mirror.common.domain.entity.EntityId;
import com.hederahashgraph.api.proto.java.AccountID;
import com.hederahashgraph.api.proto.java.Timestamp;
import com.hederahashgraph.api.proto.java.TransactionID;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

class TopicMessageTest {

    // Test serialization to JSON to verify contract with PostgreSQL listen/notify
    @ParameterizedTest
    @ValueSource(booleans = {true, false})
    void toJson(boolean nullVersion) throws Exception {
        TopicMessage topicMessage = new TopicMessage();
        topicMessage.setChunkNum(1);
        topicMessage.setChunkTotal(2);
        topicMessage.setConsensusTimestamp(1594401417000000000L);
        topicMessage.setMessage(new byte[] {1, 2, 3});
        topicMessage.setPayerAccountId(EntityId.of("0.1.1000"));
        topicMessage.setRunningHash(new byte[] {4, 5, 6});
        topicMessage.setRunningHashVersion(nullVersion ? null : 2);
        topicMessage.setSequenceNumber(1L);
        topicMessage.setTopicId(EntityId.of("0.0.1001"));
        topicMessage.setValidStartTimestamp(1594401416000000000L);

        TransactionID transactionID = TransactionID.newBuilder()
                .setAccountID(
                        AccountID.newBuilder().setShardNum(0).setRealmNum(0).setAccountNum(10))
                .setTransactionValidStart(Timestamp.newBuilder().setSeconds(20).setNanos(20))
                .setNonce(1)
                .setScheduled(true)
                .build();
        topicMessage.setInitialTransactionId(transactionID.toByteArray());
        var expectedVersion = nullVersion ? "\"running_hash_version\":null," : "\"running_hash_version\":2,";

        String json = OBJECT_MAPPER.writeValueAsString(topicMessage);
        assertThat(json)
                .isEqualTo("{" + "\"@type\":\"TopicMessage\","
                        + "\"chunk_num\":1,"
                        + "\"chunk_total\":2,"
                        + "\"consensus_timestamp\":1594401417000000000,"
                        + "\"initial_transaction_id\":\"CgQIFBAUEgIYChgBIAE=\","
                        + "\"message\":\"AQID\","
                        + "\"payer_account_id\":4294968296,"
                        + "\"running_hash\":\"BAUG\","
                        + expectedVersion
                        + "\"sequence_number\":1,"
                        + "\"topic_id\":1001,"
                        + "\"valid_start_timestamp\":1594401416000000000}");
    }
}

File: hedera-mirror-rest/__tests__/integration/topicmessage.test.js
/*
 * Copyright (C) 2024 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Range} from 'pg-range';

import * as constants from '../../constants';
import integrationDomainOps from '../integrationDomainOps';
import {setupIntegrationTest} from '../integrationUtils';
import * as utils from '../../utils';

import topicMessage from '../../topicmessage';

setupIntegrationTest();

const {ASC, DESC} = constants.orderFilterValues;
const {SEQUENCE_NUMBER, TIMESTAMP} = constants.filterKeys;
const {eq, gt, gte, lt, lte, ne} = utils.opsMap;

describe('getTopicMessageTimestampRanges', () => {
  beforeEach(async () => {
    await integrationDomainOps.loadTopicMessageLookups([
      {partition: '2022_05', sequence_number_range: '[1, 30)', timestamp_range: '[1000, 2000)', topic_id: 123},
      {partition: '2022_06', sequence_number_range: '[30, 60)', timestamp_range: '[2000, 3000)', topic_id: 123},
      {partition: '2022_07', sequence_number_range: '[60, 90)', timestamp_range: '[3000, 4000)', topic_id: 123},
    ]);
  });

  const spec = [
    {name: 'empty filters', expected: [Range(1000, 2000, '[)'), Range(2000, 3000, '[)')]},
    {name: 'empty filters descending', order: DESC, expected: [Range(2000, 3000, '[)'), Range(3000, 4000, '[)')]},
    {
      name: 'descending with largest limit for two partitions',
      limit: 31,
      order: DESC,
      expected: [Range(2000, 3000, '[)'), Range(3000, 4000, '[)')],
    },
    {
      name: 'descending with smallest limit for three partitions',
      limit: 32,
      order: DESC,
      expected: [Range(1000, 2000, '[)'), Range(2000, 3000, '[)'), Range(3000, 4000, '[)')],
    },
    {
      name: 'sequence number upper bound and ascending expect first partition',
      sequenceNumberFilters: [{key: SEQUENCE_NUMBER, operator: lt, value: 30}],
      expected: [Range(1000, 2000, '[)')],
    },
    {
      name: 'sequence number lower bound and descending expect last partition',
      sequenceNumberFilters: [{key: SEQUENCE_NUMBER, operator: gte, value: 60}],
      expected: [Range(3000, 4000, '[)')],
    },
    {
      name: 'sequence number lower bound after largest in db',
      sequenceNumberFilters: [{key: SEQUENCE_NUMBER, operator: gte, value: 90}],
    },
    {
      name: 'sequence number lower bound after largest in db and descending',
      sequenceNumberFilters: [{key: SEQUENCE_NUMBER, operator: gte, value: 90}],
      order: DESC,
    },
    {
      name: 'sequence number eq filter',
      sequenceNumberFilters: [{key: SEQUENCE_NUMBER, operator: eq, value: 5}],
      expected: [Range(1000, 2000, '[)')],
    },
    {
      name: 'multiple sequence number eq filter',
      sequenceNumberFilters: [
        {key: SEQUENCE_NUMBER, operator: eq, value: 5},
        {key: SEQUENCE_NUMBER, operator: eq, value: 6},
      ],
    },
    {
      name: 'effectively empty sequence number filter',
      sequenceNumberFilters: [
        {key: SEQUENCE_NUMBER, operator: gte, value: 10},
        {key: SEQUENCE_NUMBER, operator: lt, value: 10},
      ],
    },
    {
      name: 'effectively empty timestamp filter',
      sequenceNumberFilters: [
        {key: TIMESTAMP, operator: gte, value: 2000},
        {key: TIMESTAMP, operator: lt, value: 2000},
      ],
    },
    {
      name: 'no intersection of sequence number and timestamp filters',
      sequenceNumberFilters: [
        {key: SEQUENCE_NUMBER, operator: gte, value: 1},
        {key: SEQUENCE_NUMBER, operator: lt, value: 60},
      ],
      timestampFilters: [
        {key: TIMESTAMP, operator: gte, value: 3000},
        {key: TIMESTAMP, operator: lt, value: 4000},
      ],
    },
    {name: `topic lookup for topic id doesn't exist`, topicId: 125},
  ];

  test.each(spec)(
    '$name',
    async ({
      topicId = 123,
      sequenceNumberFilters = [],
      timestampFilters = [],
      limit = 25,
      order = ASC,
      expected = [],
    }) => {
      await expect(
        topicMessage.getTopicMessageTimestampRanges(topicId, sequenceNumberFilters, timestampFilters, limit, order)
      ).resolves.toEqual(expected);
    }
  );
});
